### ISSUE TITLE
Why I havent used Elm yet

### ISSUE THREAD

**Issue by gyzerok** [elm-guides/elm-for-js#5]
Hi, there!

I've been looking for a good functional language over JavaScript for a couple of month. And I'm trying to force myself to try Elm since it looks better in comparison to others. The only thing that prevents me from use Elm is how you do html.

Personally I've found JSX awesome. I'm familiar with it, designers in my team famaliar with it. All is good here. But Elm forces me to use highly strange syntax for this. Is there a way to use JSX with Elm? What is the reason for creating own syntax?

**Comment by gyzerok** [comment]
Also I'm curious about using Elm on the server-side.

**Comment by gyzerok** [comment]
and also i'm curious about some code-style things which feels dirty to JavaScript programmer. I do not understand the reasons behind this code-style or language grammar.
For example commas at the start of the line feels shitty. Why do not support trailig commas for better diffs?
Lets look at the example from the docs

``` elm
type alias Circle =
    { x : Float
    , y : Float
    , radius : Float
    }
```

As a JavaScript programmer I would prefer following:

``` elm
type alias Circle =
  { 
    x : Float,
    y : Float,
    radius : Float,
  }
```

Or even the CoffeeScript way

``` elm
type alias Circle =
    x : Float
    y : Float
    radius : Float
```

What I'm asking is some sort of FAQ. Mb there other syntax approaches that would feel better for JavaScript developer.

**Comment by jamischarles** [comment]
@gyzerok Thanks. This is great feedback. For context, I'm somebody learning about Elm, perhaps others can chime in as well.

>  Is there a way to use JSX with Elm? What is the reason for creating own syntax?
> JSX is syntax that is compiled to JS function calls via something like babel. Elm doesn't seem to have a templating language as far as I can tell (at least not yet). Somebody could probably write a library that does just that. Heres'a thread on the topic: https://groups.google.com/forum/#!topic/elm-discuss/naWeiwAOXzY

What Elm has is functions (basically what JSX compiles to). Personally I really like the the idea of writing markup, styling and business logic in my Elm code, but recognize that this may become unruly at a larger scale. Perhaps @evancz or @rtfeldman have more thoughts on this.

> For example commas at the start of the line feels shitty. Why do not support trailig commas for better diffs
> Elm was heavily influenced by Haskell. I imagine most of these conventions come from the Haskell community. Also, in Elm whitespace matters.

Remember that Elm compiles to JS, and is meant as a better way to do GUI programming, not just to replace JS.

Hope this is a good start. Maybe we'll get some better responses from others.

**Comment by rtfeldman** [comment]
> Is there a way to use JSX with Elm? What is the reason for creating own syntax?

As @jamischarles noted, JSX is actually the special syntax here; Elm just uses plain old functions. :smile: Elm's syntax is different than JavaScript's, of course, but that makes sense because it's a different language with different semantics - and having used both extensively, I definitely prefer Elm's syntax!

Having said that, just like how Facebook wrote a compiler from JSX to JavaScript, there's no reason someone couldn't do the same for Elm if they were so inclined. No one has, to my knowledge, but I can't think of a technical reason it wouldn't work.

> commas at the start of the line feels shitty. Why do not support trailig commas for better diffs?

There's a trade-off there, and we have actually discussed this a lot before arriving at the leading-commas style.

I used to be in favor of trailing commas because of things like better diffs, but then I did a bunch of pair programming using Elm and realized how often we would accidentally forget a trailing comma and spend time trying to debug code that looked fine but was causing an error in practice.

It's glaringly obvious when you forget a leading comma, but not when you forget a trailing one. As soon as we tried out the leading-comma style, our time spent debugging comma-related errors went straight to zero, so we adopted that style and never looked back.

As for the CoffeeScript style (with significant indentation instead of commas), that style is highly contentious. Like LISP style (and its many nested parentheses), you can make the case that it has certain objectively superior characteristics, but in practice there are a huge number of programmers who find it totally unreadable. I remember giving a conference talk about CoffeeScript once where someone in the front row got up and walked out as soon as I brought up the fact that CoffeeScript has significant indentation. As much as some of us may like it, it's sharply divisive within the community at large.

In short, leading commas improve productivity compared to trailing commas by saving you debugging time, while being far less divisive than significant indentation.

> Personally I really like the the idea of writing markup, styling and business logic in my Elm code, but recognize that this may become unruly at a larger scale. Perhaps @evancz or @rtfeldman have more thoughts on this.

Some people already do this. Personally I keep stylesheets separate mainly because I'm apprehensive about the performance implications of switching to inline styles (I have yet to hear of any serious benchmarking efforts comparing inline styles to external stylesheets), but conceptually I certainly see the appeal!

> Also I'm curious about using Elm on the server-side.

It's certainly possible, considering [`elm-test` already runs on Node](https://github.com/rtfeldman/node-elm-test). However, the support for it is definitely not first-class at the moment, as the community is strongly focused on Elm in the browser for the time being.

Over time I expect interest in Elm on the server will grow, and a library ecosystem will start growing, but I think that's a ways off for now. :smiley:

**Comment by insin** [comment]
Lack of a JSX-alike is one of the things keeping me away too. I did the [comma-first function calls/structures thing](https://gist.github.com/insin/8e72bed793772d82ca8d) for years in JavaScript; it was - and is - painful to write, read and maintain. JSX solved that for me.

**Comment by jamischarles** [comment]
I'm curious. What specifically is it that you miss from JSX?

Would you be fine with something like an es6 template string?

Something like this

``` elm

content = """
<div class="someClass">
  <a href="/about/">About</a>
</div>
"""
```

Would this cover it? Or what features are you missing specifically @insin @gyzerok ?

**Comment by pzavolinsky** [comment]
My first approach to Elm left me with similar impressions to @gyzerok and @insin in relation to the lack of a JSX-like alternative. 

Someone in my team even claimed something on lines of "the lack of JSX is a show-stopper for me".

Even though this claim might be a bit to much, I do feel that the native `elm-html` syntax makes the learning curve for Elm steeper than it needs to be (specially if you come from React).

Anyway, long story short, I implemented an [elmx pre-compiler](https://github.com/pzavolinsky/elmx) that _hopefully_ will accommodate most cases and reduce adoption friction.

I also wrote an [elmx Atom package](https://atom.io/packages/language-elmx) that includes syntax highlighting, auto-completion, linting and also compiles `.elmx` files to `.elm` every time you save.

Hope this helps!

**Comment by gouegd** [comment]
looks great, thanks for sharing @pzavolinsky  ! I'll give that a try.
Awesome that you even made an atom package for it.

**Comment by nilliams** [comment]
Nice work @pzavolinsky. I'd really like to see this as a contrib-style, officially maintained package. The lack of it in the Elm examples is a big turn-off to me and I think will be to many other developers. I feel that arguing purity or "you don't need it" is unlikely to convince people that:

```
view address model =
  div []
    [ button [ onClick address Decrement ] [ text "-" ]
    , div [] [ text (toString model) ]
    , button [ onClick address Increment ] [ text "+" ]
    ]
```

is more readable/writable/manageable than something like (pseudo code ... I've not familiarised myself with elmx yet):

```
view address model =
  <div>
    <button onClick={ address Decrement }>-</button>
    <div>{ toString model }</div>
    <button onClick={ address Increment }>+</button>
  </div>
```

It'd be great to see Elm embrace JSX-like syntax, as it has clearly helped React adoption, the fact that it's the norm over there, over writing render methods in the pure-JS, or CoffeeScript style should be telling.

Also agree the comma-first style is hostile. It would be sad to see it become the standard in Elm. I was just listening to [a podcast](http://thewebplatformpodcast.com/76-the-elm-programming-language) where an `elm-format` linter tool (like `gofmt`) was discussed -- I really hope it doesn't enforce comma-first.

**Comment by rtfeldman** [comment]
> It'd be great to see Elm embrace JSX-like syntax

There's a "simplicity vs familiarity" trade-off here: plain old vanilla functions are simpler, and although introducing an entire templating syntax layer can add some familiarity, it also adds a lot of complexity.

Simplicity is one of Elm's strengths, and maintaining that simplicity requires carefully evaluating tradeoffs like this.

I think Evan made the right call in going with vanilla functions, and I don't think that's likely to change. :smile:

**Comment by gyzerok** [comment]
Since my first post I've done some development in Elm. I do not see any simplicity trade-off between JSX and functions @rtfeldman said about, but I've found myself almost comfortable with functions.

I guess this issue may be closed :)

**Comment by roman** [comment]
I still consider, if we want Elm "to be mainstream", we need to accommodate jsx somehow, it is a dealbreaker for a lot of developers that not into functional from the getgo :-(

**Comment by mgold** [comment]
Someone could write a tool to convert jsx into Elm html lists.

**Comment by pzavolinsky** [comment]
Hi @mgold , @roman , I had similar concerns, that's why I wrote:

> [Elmx](https://github.com/pzavolinsky/elmx) - A tiny precompiler that takes an Elm program with embedded HTML and desugars the HTML into elm-html syntax. Elmx is to Elm what React's JSX is to Javascript

(granted, is an external tool and not part of a the official `elm-make` compiler, but it should do the trick)

Cheers!

**Comment by ghost** [comment]
While i've read through all this, I'm curious if vanilla elm is really so hard to adapt. I mean, if you've learned the basics, how function calls work, whats so hard about:

```
someTagName [ list of attributes ] [ list of children ]
```

It's short, clean and requires less typing then HTML while its still clear to see what belongs to what. If your view gets more complex, it surely will become less readable. I personally feel the same way about JSX. So I do the same as in JSX: I go the simple way and split up my view into several, more easy to read and maintainable functions. Problem solved.

Don't get me wrong, I like JSX in first place and the things it has to offer, but I don't see any sense in making elm something what it not is. Elm is elm and thats fine as JSX is what it is.

**Comment by nilliams** [comment]
Sure it's not _that_ hard to adapt, other than having to remember the empty `[ ]` when you have no attributes, it just has 'worse affordance' relative to the DOM it translates to.

By that I mean, I still have to mentally map it to the DOM it will be translated to, so why make me work harder to do that mental mapping.

(Also, some people like me find it ugly, appreciate that's opinion though).

> Don't get me wrong, I like JSX in first place and the things it has to offer, but I don't see any sense in making elm something what it not is. Elm is elm and thats fine as JSX is what it is.

That's a false dichotomy though, as you don't have to 'change Elm' at all to introduce a 'contrib' support for something like an [elmx](https://github.com/pzavolinsky/elmx), which would be entirely optional just like JSX is.

**Comment by develerltd** [comment]
How about this format of display?

```css
div[] [
  button[onClick=Decrement] [
    text "-"
  ]
  , div[] [
    text (toString model)
  ]
  , button[onClick=Increment] [
    text "+"
  ]
]
```

I think it would be clearer if in this circumstance you could wrap the second arguments with "{}" instead of [] when there will be more than one level of components?

eg:

```css
div[] [
  button[onClick=Decrement] [
    text "-"
  ]
  , div[] {
    h1[class="my-header"] [
        text "Hello World"
    ]
  }
  , button[onClick=Increment] [
    text "+"
  ]
]
```

**Comment by variousauthors** [comment]
Template engines were great for migrating a codebase. If we were all still using templating engines this would not be a problem: I would copy/paste all my haml templates from my old codebase into my elm codebase, and elm's package ecosystem would provide a template engine for me.

Using markup inline is a step forward, let's not also take two steps back. If we are all composing out HTML inline, we should be doing it in language agnostic way. If, going forward, we are all (purescript, elm, react, reasonml) going to use the "write your markup in your components" mode in preference to the old templating engines, then we should expect to be able to share our JSX in the same way as we were able to share our templates.

That's my feeling.
